internal/testintegration/setup.go

```diff
@@ -110,85 +107,140 @@ func EnvTestSetup() (func(), error) {
 		},
 	}
 
-	teardownFunc := func() {
-		cancel()
-		err := testEnv.Stop()
-		if err != nil {
-			Log.Error(err, "unable to stop envtest environment %v")
-		}
-	}
-
+	// Start the testenv
 	cfg, err := testEnv.Start()
+	th := &EnvTestHarness{
+		testEnvCtx:    ctx,
+		testEnv:       testEnv,
+		testEnvCancel: cancel,
+		cfg:           cfg,
+		cancel:        func() {},
+	}
 	if err != nil {
-		return nil, fmt.Errorf("unable to start kuberenetes envtest %v", err)
+		return th, fmt.Errorf("unable to start kuberenetes envtest %v", err)
 	}
 
-	s := scheme.Scheme
-	controller.InitScheme(s)
-
-	Client, err = client.New(cfg, client.Options{Scheme: s})
+	// Initialize rest client configuration
+	th.s = scheme.Scheme
+	controller.InitScheme(th.s)
+	cl, err := client.New(cfg, client.Options{Scheme: th.s})
 	if err != nil {
-		return teardownFunc, fmt.Errorf("unable to start kuberenetes envtest %v", err)
-	}
-	if Client == nil {
-		return teardownFunc, fmt.Errorf("unable to start kuberenetes envtest %v", err)
+		return th, fmt.Errorf("unable to to create client %v", err)
 	}
+	th.Client = cl
 
-	mgrCancelFunc, err := startManager(ctx, cfg, s, workload.DefaultProxyImage)
+	// Start the controller-runtime manager
+	err = th.StartManager(workload.DefaultProxyImage)
 	if err != nil {
-		return teardownFunc, fmt.Errorf("unable to start kuberenetes envtest %v", err)
+		return th, fmt.Errorf("unable to start kuberenetes envtest %v", err)
 	}
 
-	RestartManager = func(defaultProxyImage string) error {
-		mgrCancelFunc()
-		mgrCancelFunc, err = startManager(ctx, cfg, s, defaultProxyImage)
-		return err
+	return th, nil
+}
+
+// EnvTestHarness enables integration tests to control the lifecycle of the
+// operator's controller-manager.
+type EnvTestHarness struct {
+
+	// Mgr is the manager
+	Mgr ctrl.Manager
+
+	// Client is the kubernetes client.
+	Client client.Client
+
+	// The actual EnvTest environment
+	testEnv *envtest.Environment
+
+	// testEnvCancel is the context cancel function for the testEnv
+	testEnvCancel context.CancelFunc
+
+	// cancel is the context cancel function for the manager
+	cancel context.CancelFunc
+
+	// stopped channel is closed when the manager actually stops
+	stopped chan int
+
+	// cfg is the client configuration from envtest
+	cfg *rest.Config
+
+	// s is the client scheme
+	s          *runtime.Scheme
+	testEnvCtx context.Context
+}
+
+// Teardown closes the TestEnv environment at the end of the testcase.
+func (h *EnvTestHarness) Teardown() {
+	h.testEnvCancel()
+	err := h.testEnv.Stop()
+	if err != nil {
+		Log.Error(err, "unable to stop envtest environment %v")
 	}
+}
+
+// StopManager stops the controller manager and waits for it to exit, returning an
+// error if the controller manager does not stop within 1 minute.
+func (h *EnvTestHarness) StopManager() error {
+	h.cancel()
+
+	// reassign cancel to do nothing, now that we've already
+	// called it.
+	h.cancel = func() {}
```

Comment:
Canceling a CancelFunc repeatedly is actually totally fine.

https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/context/context.go;l=463

Reply:
Fixed. I removed this line.